batch.go

```diff
@@ -991,3 +1004,121 @@ func (batch *Batch) Equal(other Batcher) bool {
+	for i := 0; i < len(b.Entries); i++ {
+		// TODO(adam): Should we remove this based on checking the last element is
+		// debit/credit and sums to all the other elements (which are mutually exclusive to
+		// the last record being debit or credit)?
+		if strings.EqualFold(b.Entries[i].IndividualName, "OFFSET") {
+			// fixup BatchControl records for our conditional after this for loop
+			if b.Entries[i].TransactionCode == CheckingCredit || b.Entries[i].TransactionCode == SavingsCredit {
+				b.Control.TotalCreditEntryDollarAmount = 0
+			} else {
+				b.Control.TotalDebitEntryDollarAmount = 0
+			}
+			// remove the EntryDetail
+			b.Entries = append(b.Entries[:i], b.Entries[i+i:]...)
+			addOffset = false // we're replacing an EntryDetail, not adding
+			i++
+		}
+	}
+	if b.Control.TotalDebitEntryDollarAmount > 0 && b.Control.TotalCreditEntryDollarAmount > 0 {
+		return fmt.Errorf("WithOffset: batch must contain only debits=%d or credits=%d", b.Control.TotalDebitEntryDollarAmount, b.Control.TotalCreditEntryDollarAmount)
+	}
+	if utf8.RuneCountInString(b.offset.RoutingNumber) != 9 {
+		return fmt.Errorf("WithOffset: invalid routing number %s", b.offset.RoutingNumber)
+	}
+
+	ed := NewEntryDetail()
+	if code, err := b.determineTransactionCode(); err != nil {
+		return err
+	} else {
+		ed.TransactionCode = code
+	}
+	ed.RDFIIdentification = b.offset.RoutingNumber[:8]
+	ed.CheckDigit = b.offset.RoutingNumber[8:9]
+	ed.DFIAccountNumber = b.offset.AccountNumber
+	ed.IdentificationNumber = "" // TODO(adam): what to put here?
+	ed.IndividualName = "OFFSET"
+	ed.DiscretionaryData = b.offset.Description
+	ed.TraceNumber = createTraceNumber(b.offset.RoutingNumber)
+
+	if b.Control.TotalDebitEntryDollarAmount > 0 {
+		ed.Amount = b.Control.TotalDebitEntryDollarAmount
+		b.Control.TotalCreditEntryDollarAmount = ed.Amount
+	} else {
+		ed.Amount = b.Control.TotalCreditEntryDollarAmount
+		b.Control.TotalDebitEntryDollarAmount = ed.Amount
+	}
+	if ed.Category == "" {
+		ed.Category = CategoryForward
+	}
+	b.AddEntry(ed)
+	if addOffset {
+		b.Control.EntryAddendaCount += 1
+	}
+	b.Header.ServiceClassCode = MixedDebitsAndCredits
+	b.Control.EntryHash = b.calculateEntryHash()
+	return nil
+}
+
+func (b *Batch) determineTransactionCode() (int, error) {
+	if b == nil || b.offset == nil {
+		return 0, errors.New("nil Batch or nil Batch.offset")
+	}
+	switch b.offset.AccountType {
+	case OffsetChecking:
+		if b.Control.TotalDebitEntryDollarAmount > 0 {
+			return CheckingCredit, nil
+		} else {
+			return CheckingDebit, nil
+		}
+	case OffsetSavings:
+		if b.Control.TotalDebitEntryDollarAmount > 0 {
+			return SavingsCredit, nil
+		} else {
+			return SavingsDebit, nil
+		}
+	default:
+		return 0, fmt.Errorf("offset: unknown account type %v", b.offset.AccountType)
+	}
+}
+
+// aba8 returns the first 8 digits of an ABA routing number.
+// If the input is invalid then an empty string is returned.
+func aba8(rtn string) string {
```

Comment:
Do we not have something that does this already?

Reply:
This was copied from paygate, but I didn't see a function like this in the ACH project.